src/api.js

```diff
@@ -0,0 +1,146 @@
+ */
+export async function getCoffeeShops() {
+    let token;
+    let retries = 0;
+
+    while (retries < MAX_RETRIES) {
+        try {
+            if (!token) {
+                const tokenResult = await getToken();
+                token = tokenResult.token;
+                console.log('[AUTH] Token obtained');
+            }
+            console.log('[SHOPS] Fetching for shops.');
+            return await fetchWithRetry(
+                `${API_URL}/v1/coffee_shops`,
+                { method: 'GET' },
+                token,
+                '[SHOPS]',
+            );
+        } catch (err) {
+            if (err.needsNewToken) {
+                token = null;
+                continue;
+            }
+
+            retries++;
+            if (retries >= MAX_RETRIES) {
+                throw new Error(
+                    `[SHOPS] Failed to get coffee shops after ${MAX_RETRIES} attempts: ${err.message}`,
+                );
+            }
+
+            const delay = retries * RETRY_DELAY;
+            console.error(
+                `[SHOPS] Error: ${err.message}. Retrying in ${delay}ms (${retries}/${MAX_RETRIES})`,
+            );
+            await sleep(delay);
+        }
+    }
+}
+
+/**
+ * Get an authentication token from the API
+ * @returns {Promise<{token: string}>} - Object containing the token
+ */
+export async function getToken() {
+    try {
+        return await fetchWithRetry(
+            `${API_URL}/v1/tokens`,
+            { method: 'POST' },
+            null,
+            '[AUTH]',
+        );
+    } catch (err) {
+        throw new Error(`Failed to get authentication token: ${err.message}`);
+    }
+}
+
+/**
+ * Wrapper for fetch with retry logic, timeouts, and token handling
+ * @param {string} url - API endpoint URL
+ * @param {Object} options - Fetch options
+ * @param {string} [token] - Authentication token
+ * @param {string} logPrefix - Prefix for log messages
+ * @returns {Promise<any>} - Parsed JSON response
+ */
+async function fetchWithRetry(url, options = {}, token = null, logPrefix = '') {
+    let retries = 0;
+    let currentToken = token;
+
+    while (retries < MAX_RETRIES) {
+        const controller = new AbortController();
+        const timeoutId = setTimeout(() => controller.abort(), REQUEST_TIMEOUT);
+
+        try {
+            const urlWithToken = currentToken
+                ? `${url}?token=${currentToken}`
+                : url;
+
+            const response = await fetch(urlWithToken, {
+                'Content-Type': 'application/json',
+                Accept: 'application/json',
+                ...options,
+                signal: controller.signal,
+            });
+
+            clearTimeout(timeoutId);
+
+            if (response.status === 401) {
+                console.log(`${logPrefix} Token expired, getting new token`);
+                throw { status: 401, needsNewToken: true };
+            }
+
+            if (response.status === 503 || response.status === 504) {
```

Comment:
The intention is not very clear for these "magic numbers". I can't say off the top of my head what 503 and 504 means when I'm reading the code for the first time and why you're handling these errors the way you do.

How would you make the meaning of the numbers more obvious?

Reply:
Yeah, that is what I'm also thinking about